### ISSUE TITLE
UnusedPrivateMethod and callables

### ISSUE THREAD

**Issue by yitznewton** [phpmd/phpmd#134]
PHPMD is not detecting use as a callable, e.g.

``` php
class Mapper
{
    private function foo($item)
    {
        return $item;
    }

    public function bar()
    {
        return array_map(array($this, 'foo'), array('baz'));
    }
}
```

See https://github.com/yitznewton/phpmd/pull/1 for failing test

I looked into pdepend, but I did not see any Source classes that dealt with this construct.

**Comment by AbcAeffchen** [comment]
:+1:

**Comment by leafpeak** [comment]
+1 For this!

**Comment by tarlepp** [comment]
And seems like this is still valid issue - any updates / ETA for actual fix ?

**Comment by dmchale** [comment]
+1 looking for fix. I reported this to codacy as a bug since I was getting a false positive in one of my projects, and they pointed me here since their reports simply return what PHPMD detects. It seems that array_map itself is the problem, since it fails whether the function being called is a global function or a class-based function.

**Comment by kylekatarnls** [comment]
I'm not sure we could (and should) fix this. I highly recommend the following syntax:

```php
return array_map(function ($word) {
  return $this->foo($word);
}, array('baz'));
```

I know, it's more verbose but it has many benefits:
- I can click on ->foo with my favorite IDE and reach the function definition, even modern ones can't find the declaration for `[$this, 'foo']` this syntax defeat them as much PHPMD.
- The number of parameters passed are controlled: imagine the following: some developer find `foo($item)` and think "the second parameter is available for an other option, so he add some option: `foo($item, $suffix)` and return `$item.$suffix` this will make it very hard to find why `bar()` suddently returns `baz0` instead of `baz`, it's because array_map pass the key/index as the second parameter. If you use the closure, it won't so you can consciously use the second parameter and pass it to foo or not. And no matter the change you made on the inner method or the function/method that take closures as parameters.
- If you allow this syntax, you open the door to patterns like: `array($this, 'foo'.$anything)` obviously no static analyzer could possibly handle this.

So we would have to detect and re-compose static string, detect that it can refer to an existing method and is used as a callable by the method/function calling.

Let's be realistic, I highly doubt we could have a correct implementation of this feature (working without any false positives/negatives warnings).

And as a tool recommending best practices, if you use PHPMD, you probably care about using the closure option. Anyone facing this issue should rather consider it.

**Comment by tarlepp** [comment]
Yeah that is one way to avoid this error, but eg. psalm detects this correctly - https://psalm.dev/r/39a23b2e83

**Comment by kylekatarnls** [comment]
But it does not with the example of dynamic ones:
https://psalm.dev/r/b86e124d51

(Even when this is actually static like in this example). And we won't implementing it just because an other tool do.

Don't misunderstand me. I'm not saying you should use closure as a work-around to avoid this error, I say you should avoid this pattern for the 4 reasons given above. Referring methods using the array-syntax is a bad practice anyway we handle it or not. So I just think supporting this not-so-good syntax does not worth the effort and risk.

**Comment by tarlepp** [comment]
Yeah I see your point there. What about using detailed error message with this to show people how to avoid this one?

**Comment by kylekatarnls** [comment]
Fixed by #1068
